libvast/vast/arrow_extension_types.hpp

```diff
@@ -158,24 +173,350 @@ class pattern_extension_type : public arrow::ExtensionType {
+// in terms of indexing for `arrow::Array`.
+// For this machinery to work, we need to create a separate struct extending
+// `arrow::ExtensionArray` for every custom extension type and provide a
+// proper implementation for `::TypeClass`. See `pattern_array` etc.
+
+template <>
+struct sum_type_access<arrow::Array> final {
+  template <class T>
+  struct is_extension_array
+    : std::integral_constant<bool, T::TypeClass::type_id
+                                     == arrow::ExtensionType::type_id> {};
+  template <class Types>
+  struct tl_map_array_to_type;
+  template <class... Ts>
+  struct tl_map_array_to_type<detail::type_list<Ts...>> {
+    using type = detail::type_list<typename Ts::TypeClass...>;
+  };
+  using types = detail::type_list<
+    arrow::NullArray, arrow::BooleanArray, arrow::Int64Array,
+    arrow::UInt64Array, arrow::DoubleArray, arrow::DurationArray,
+    arrow::StringArray, arrow::TimestampArray, arrow::MapArray,
+    arrow::ListArray, arrow::StructArray, vast::address_array,
+    vast::pattern_array, vast::enum_array, vast::subnet_array>;
+  using data_types = typename tl_map_array_to_type<types>::type;
+  using extension_types
+    = detail::tl_filter_t<data_types, arrow::is_extension_type>;
+
+  using type0 = detail::tl_head_t<types>;
+  static constexpr bool specialized = true;
+
+  template <class T, int Index>
+  static bool is(const arrow::Array& arr, sum_type_token<T, Index>) {
+    if (arr.type_id() != T::TypeClass::type_id)
+      return false;
+    if constexpr (is_extension_array<T>::value)
+      return static_cast<const arrow::ExtensionType&>(*arr.type())
+               .extension_name()
+             == T::TypeClass::vast_id;
+    return true;
+  }
+
+  template <class T, int Index>
+  static const T& get(const arrow::Array& arr, sum_type_token<T, Index>) {
+    return static_cast<const T&>(arr);
+  }
+
+  template <class T, int Index>
+  static T& get(arrow::Array& arr, sum_type_token<T, Index>) {
+    return static_cast<T&>(arr);
+  }
+
+  template <class T, int Index>
+  static const T*
+  get_if(const arrow::Array* arr, sum_type_token<T, Index> token) {
+    if (arr && is(*arr, token))
+      return &get(*arr, token);
+    return nullptr;
+  }
+
+  template <class T, int Index>
+  static T* get_if(arrow::Array* arr, sum_type_token<T, Index> token) {
+    if (arr && is(*arr, token))
+      return &get(*arr, token);
+    return nullptr;
+  }
+
+  static int index_from_type(const arrow::DataType& x) noexcept;
+
+  template <class Result, class Visitor, class... Args>
+  static auto apply(const arrow::Array& x, Visitor&& v, Args&&... xs)
+    -> Result {
+    // A dispatch table that maps variant type index to dispatch function for
+    // the concrete type.
+    static constexpr auto table = []<class... Ts, int... Indices>(
+      detail::type_list<Ts...>,
+      std::integer_sequence<int, Indices...>) noexcept {
+      return std::array{
+        // decay to function pointer
```

Comment:
This comment only applies to the unary + operator below, which makes this more confusing rather than helping the reader.

The array maps a variant index to a function pointer that does the actual visit dispatch.